controllers/imageclusterinstall_controller.go

```diff
@@ -521,27 +524,55 @@ func (r *ImageClusterInstallReconciler) SetupWithManager(mgr ctrl.Manager) error
 		Complete(r)
 }
 
-func (r *ImageClusterInstallReconciler) setBMHImage(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, url string) error {
+func (r *ImageClusterInstallReconciler) getBmhDataImage(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost) *bmh_v1alpha1.DataImage {
+	if bmh == nil {
+		return nil
+	}
+	dataImage := bmh_v1alpha1.DataImage{}
+	key := client.ObjectKey{
+		Name:      bmh.Name,
+		Namespace: bmh.Namespace,
+	}
+	err := r.Get(ctx, key, &dataImage)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+
+	return &dataImage
+}
+
+func (r *ImageClusterInstallReconciler) setBMHDataImage(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, url string) error {
+	dataImage := r.getBmhDataImage(ctx, bmh)
+
+	if dataImage == nil {
+		r.Log.Info("creating new dataImage", "name", bmh.Name, "namespace", bmh.Namespace)
+		dataImage = &bmh_v1alpha1.DataImage{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      bmh.Name,
+				Namespace: bmh.Namespace,
+			},
+			Spec: bmh_v1alpha1.DataImageSpec{
+				URL: url,
+			},
+		}
+		err := controllerutil.SetControllerReference(bmh, dataImage, r.Client.Scheme())
+		if err != nil {
+			return fmt.Errorf("failed to set controller reference for dataImage due to %w", err)
+		}
+
+		err = r.Create(ctx, dataImage)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve dataImage due to %w", err)
+		}
+	}
+
 	patch := client.MergeFrom(bmh.DeepCopy())
 
 	dirty := false
 	if !bmh.Spec.Online {
 		bmh.Spec.Online = true
 		dirty = true
 	}
-	if bmh.Spec.Image == nil {
-		bmh.Spec.Image = &bmh_v1alpha1.Image{}
-		dirty = true
-	}
-	if bmh.Spec.Image.URL != url {
-		bmh.Spec.Image.URL = url
-		dirty = true
-	}
-	liveIso := "live-iso"
-	if bmh.Spec.Image.DiskFormat == nil || *bmh.Spec.Image.DiskFormat != liveIso {
-		bmh.Spec.Image.DiskFormat = &liveIso
```

Comment:
How does this change affect the BMH status?
Will the BMH move to provisioned or stay available?